python/lsst/ts/mtdome/mtdome_csc.py

```diff
@@ -511,48 +555,64 @@ async def write_then_read_reply(
+                )
                 raise
 
             disabled_commands: set[str] = set()
             if self.simulation_mode == ValidSimulationMode.NORMAL_OPERATIONS:
                 disabled_commands = COMMANDS_DISABLED_FOR_COMMISSIONING
 
             if command not in disabled_commands:
-                self.writer.write(st.encode() + b"\r\n")
-                await self.writer.drain()
                 try:
-                    assert self.reader is not None
-                    read_bytes = await asyncio.wait_for(
-                        self.reader.readuntil(b"\r\n"), timeout=_TIMEOUT
+                    await self.client.write_json(data=command_dict)
+                    data = await asyncio.wait_for(
+                        self.client.read_json(), timeout=_TIMEOUT
                     )
                 except (
                     asyncio.exceptions.IncompleteReadError,
                     asyncio.exceptions.TimeoutError,
-                    ConnectionResetError,
+                    ConnectionError,
                     AssertionError,
                 ) as e:
                     await self.fault(
-                        code=3, report=f"Error reading reply to command {st}: {e}"
+                        code=3,
+                        report=f"Error reading reply to command {command_dict}: {e}.",
                     )
                     raise
-                data = encoding_tools.decode(read_bytes.decode())
+                received_command_id = data["commandId"]
+                if received_command_id in self.commands_without_reply:
+                    self.commands_without_reply.pop(received_command_id)
+                else:
+                    self.log.warning(
+                        f"Ignoring unknown commandId {received_command_id}."
+                    )
             else:
                 data = REPLY_DATA_FOR_DISABLED_COMMANDS
             response = data["response"]
 
             if response != ResponseCode.OK:
                 self.log.error(f"Received ERROR {data}.")
-                if response == ResponseCode.COMMAND_REJECTED:
-                    raise ValueError(f"The command {command} was rejected.")
-                elif response == ResponseCode.UNSUPPORTED_COMMAND:
-                    raise KeyError(f"The command {command} is unsupported.")
+                match response:
+                    case ResponseCode.INCORRECT_PARAMETERS:
+                        error_message = f"{command=} has incorrect parameters."
+                    case ResponseCode.INCORRECT_SOURCE:
+                        error_message = f"{command=} was sent from an incorrect source."
+                    case ResponseCode.INCORRECT_STATE:
+                        error_message = f"{command=} was sent for an incorrect state."
+                    case _:
+                        error_message = f"{command=} is not supported."
+                raise ValueError(error_message)
```

Comment:
I gently suggest a case or dict of suffixes:
```
error_suffix = {
    ResponseCode.INCORRECT_PARAMETERS: "has incorrect parameters.",
    ResponseCode.INCORRECT_SOURCE:, "was sent from an incorrect source.",
    ...
}.get(respose, "is not supported.")
raise ValueError(f"{command=} {error_suffix"}"
```
Case works fine as well; the point is to reduce duplication of f"{command=}".